### DISCUSSION TITLE
Leaf Forum

### DISCUSSION THREAD

**Discussion by mychidarko** [leafsphp/leaf#37] [General]
# 👋🏽 Welcome to the forum

Hello! Welcome to the discussion forum for Leaf. This is a great place to introduce a project you’re working on, get help with something, discuss a plugin or particular technical topic, or just get to know other users in the community.  We hope that you:

  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Talk about challenges you've had using Leaf, and solutions if any.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

In case of questions and challenges, in order to receive the best answers to your questions please follow these simple guidelines.

- Be descriptive about your question or problem and be sure to include any errors. Single line topics will rarely receive answers and do not just post a screenshot.

- Provide examples of the code in question and use syntax highlighting.

- Break down the issue in your code and provide a link to an example that replicates it (if applicable).

Following these few steps will greatly improve your chances of getting a good quality answer to your issue. Remember, everyone is here to share ideas and to help out using their own free time. Making an effort to craft a well-thought-out topic pays tribute to this.

**Comment by propeller-aaron** [comment]
Thanks for making this! I have been looking for an alternative to Silex for a few years now. I have tried other frameworks like Lumen, Laravel & Slim, but they were always a pain to setup with my VM's, and usually forced you to adhere to a console generated bundle.

**Comment by jamie-syon** [comment]
I'm currently looking for a framework to replace our in-house built MVC framework which we have spent quite some time building over the years.

I just don't "get" the major frameworks and to me I want something simple which I can add to if needed and now I come across Leaf.  Wow.  I think you've done a great job!  I'll be getting this set up tomorrow and working through it.

Thanks!

**Comment by ivan-au** [comment]
G'day @darko-mychi 👋,

Loving Leaf! I have one suggestion to improve cookie management: Add the [SameSite](https://web.dev/samesite-cookies-explained/) option, setting the default to "Lax". SameSite cookies are well supported by modern browsers and remove the need to implement CSRF controls to forms, particularly to login forms.

Cheers!

**Reply by mychidarko** [reply]
Hi @ivan-au, thanks for your support. I'll check out what you suggested😇

**Comment by wfsdaj** [comment]
cannot access:
[https://skeleton.leafphp.dev/docs/introduction/](https://skeleton.leafphp.dev/docs/introduction/)

**Reply by mychidarko** [reply]
Hey @wfsdaj, this has been fixed already. The update will be pushed soon. Use [this](https://skeleton.leafphp.dev/docs/) link instead.

**Comment by raldincasidar-studio** [comment]
Hello @mychidarko , I appreciate leaf php so much, im having an error with my code in database module

```php
$order_info = db()->query("SELECT ord.id as order_id, ord.*, stck.* 
                    FROM customer_order ord 
                    LEFT JOIN stocks stck ON ord.stock_id = stck.id 
                    WHERE ord.id = $id;")->first();

$payments = db()->select('payments')->all();
```

Im having an error on PDO:

`PDOException: SQLSTATE[HY000]: General error: 2014 Cannot execute queries while there are pending result sets. Consider unsetting the previous PDOStatement or calling PDOStatement::closeCursor() in file C:\xampp\htdocs\ej-marketing-backend\vendor\leafs\db\src\Db\Core.php on line 318`

How to fix this?

**Reply by mychidarko** [reply]
Hi @raldincasidar-studio, are these the only queries you have on the page you're executing?

**Reply by raldincasidar-studio** [reply]
hello @mychidarko , thank you for notice, 

to answer your question, yes, that is the only query that have error when i'm executing.

However, I manage to fix this problem and I thought this is a issue from the internal code of LeafPHP but I found out it was my fault, the reason for causing the error is the semi-colon at the end of the of the first query.

That `;` inside the `query("")` caused the error and when I removed it, the error does not occur anymore. 

I appreciate your project @mychidarko and I wanted to contribute sooner with my ideas to improve LeafPHP. I hope this project will be maintained forever and I'm looking forward with sponsoring this project after my work.

**Reply by mychidarko** [reply]
Thanks @raldincasidar-studio. Your contributions would be very much appreciated. Looking forward to the outcome of your project 😊

**Comment by toddy15** [comment]
Hi @mychidarko,

first of all, thanks for creating leaf! This is an amazing new framework / module system. I've a couple of questions for contributing code, though:

1. You've mentioned in my [PR](https://github.com/leafsphp/form/pull/3#issuecomment-1250994117) that there's a style set for PHP CS Fixer in the main leaf project. Would you be willing to accept PRs for modules which add the same PHP CS Fixer setup (and reformats of the source code) for all leaf modules?
2. Which PHP version is the minimum requirement for leaf? I've guessed that it might be at least PHP 7.4, but I couldn't find any detailed information about this.

Regards,
Tobias

**Reply by mychidarko** [reply]
Thanks, @toddy15.

1. Your PRs are all welcome. They'll be a big help actually, I'll really appreciate that.
2. Well, Leaf was written to go as far back as PHP 7.1 and 7.2, I still see some installs from people using PHP 7.2 and 7.3. So I guess that's our minimum version requirement.

**Comment by scorpionvernom** [comment]
![Screenshot_20230228-194647.png](https://user-images.githubusercontent.com/69682771/222230260-84b55ecc-d6ad-4608-8f9c-5c107f1e5f90.png)

Please how can I solve this issue?

I'm just a user 🙏

**Reply by scorpionvernom** [reply]
I have tried 3 different cpanels from 3 different hosts

No way!😪

**Reply by mychidarko** [reply]
@scorpionvernom I see leafmailer.pw here, that's not associated to the Leaf framework

**Reply by mychidarko** [reply]
@scorpionvernom I saw you're still stuck on this, check the leafmailer.pew site for support

**Comment by kczx3** [comment]
I can't seem to get the 'app.down' config to be acknowledged.
```php
<?php

use Leaf\App;

require __DIR__ . '/vendor/autoload.php';

$app = new App([
    'app.down' => true
]);

// these don't work either
// $app->config('app.down', true);
// $app->configure([
//     'app.down' => true
// ]);

$app->get('/', function () use ($app) {
    $app->response()->json(Leaf\Config::get('app'));
});

$app->run();
```
Just starting to look at this framework and I really like it on the surface.  I've had issues with updating it as new versions come out though.  It seems that `leafs/leaf` doesn't require any specific versions of its other `leafs` dependencies.    This makes it impossible to commit the composer.lock file to version control if we choose to not use the CLI.  Any plans to remedy that or suggestions?

**Reply by mychidarko** [reply]
Hi @kczx3 for the issue you're having, what version of `leafs/leaf` do you have installed?

**Reply by mychidarko** [reply]
`leafs/leaf` uses `*` for the version constraint so you can install any precise version of a module you need in your app without running into conflicts. If you want to upgrade any module, you can go ahead, and that update will then be committed to your project's lock file.

**Reply by kczx3** [reply]
> Hi @kczx3 for the issue you're having, what version of `leafs/leaf` do you have installed?

I had just deleted `vendor` and `composer.lock` and re-installed the latest version - `3.3.0`.

> `leafs/leaf` uses `*` for the version constraint so you can install any precise version of a module you need in your app without running into conflicts. If you want to upgrade any module, you can go ahead, and that update will then be committed to your project's lock file.

I see what you mean but I didn't install any modules other than `leafs/leaf`.  At the time, `Anchor` was version `1.2.1` and didn't have `toBool`.  I noticed the above issue with "app.down" so the first thing I tried was `composer update leafs/leaf` however that didn't install an updated version of `leafs/Anchor`, which thus far in my project, has only been required by `leafs/leaf`.  It doesn't really make sense for me to install `leafs/anchor` just to ensure I have a version that works with the version of `leafs/leaf` when I'd expect `leafs/leaf` to handle its own dependency.  Although I guess you could say its a "peer dependency" like how NodeJS does it.

**Reply by Kristories** [reply]
@scorpionvernom create a new thread. Your comments everywhere.

**Reply by kczx3** [reply]
@mychidarko just checking if you needed any further information from me on the issue with `add.down` not working.  Thanks for your time 😄

**Reply by mychidarko** [reply]
Hi @kczx3, sorry about the delayed responses. I'll get back to you in a bit

**Reply by mychidarko** [reply]
After checking through, I noticed this is an issue with the config parser. I'm converting this to an issue so it'll fixed and updated.

**Reply by mychidarko** [reply]
https://github.com/leafsphp/leaf/issues/184

**Reply by fadrian06** [reply]
> ...so the first thing I tried was `composer update leafs/leaf` however that didn't install an updated version of `leafs/Anchor`, ...

Have you tried `composer update -W`?
to update dependencies of dependencies?

**Comment by kevyworks** [comment]
Hi, can you make samples for `Leaf\Veins`? like `{includes/function}` I'm currently using `--basic` app in leaf.

I would also suggest putting more efforts on commenting your codes.

I love laravel but seeing Laravel's blade/model/etc libraries & codes when you want lightest codebase of leafphp is not good IMHO, rather use laravel correct? maybe make your own Model Class atleast? still that's just me 😅

Cheers!

**Reply by mychidarko** [reply]
I'm still making samples for different projects in the ecosystem. Laravel-based libraries are just options for those who are used to those platforms. Laravel models only ship with Leaf MVC setups by default.

**Reply by mychidarko** [reply]
<img width="254" alt="image" src="https://user-images.githubusercontent.com/26604242/235090867-1d8323b9-b047-4577-ae4c-eff1f164b447.png">
<img width="766" alt="image" src="https://user-images.githubusercontent.com/26604242/235091388-df33ce60-2fd1-410d-87aa-1f893ad31230.png">

As you can see here, we give you different options, you can choose anything you want based on your own preference

**Comment by PiscesDavid** [comment]
Hello,
I confess I'm new to Leaf.
I'm trying to simply create a controller (using php leaf g:controller dsos) which worked.

```
namespace App\Controllers;

class DsosController extends Controller
{
    public function index()
    {
        response()->json([
            'message' => 'DsosController@index output'
        ]);
    }
}
```
Now, I want to create a route to this controller as htts://myserver/Dsos.
So, in /app/routes/_app.php, I added this line:
`app()->get('/dsos', 'DsosController@index');`
But it does not work.
Any help will be appreciated for such simple exercise,
Thanks

Cheers

**Reply by mychidarko** [reply]
Hey @PiscesDavid, can I see your `_app.php` file

**Reply by PiscesDavid** [reply]
Sure, here it is:
```
<?php

app()->get("/", function () {
    response()->json(["message" => "Congrats!! You're on Leaf API"]);
});

app()->get("/app", function () {
    // app() returns $app
    response()->json(app()->routes());
});

app()->get('/dsos', 'DsosController@index');
```
Many thanks for the help

**Reply by mychidarko** [reply]
This looks good, can you seed your `routes/index.php` file

**Reply by PiscesDavid** [reply]
Here it is:
```
<?php

/*
|--------------------------------------------------------------------------
| Set up 404 handler
|--------------------------------------------------------------------------
|
| Create a handler for 404 errors
|
*/
app()->set404(function () {
	response()->json("Resource not found", 404, true);
});

/*
|--------------------------------------------------------------------------
| Set up 500 handler
|--------------------------------------------------------------------------
|
| Create a handler for error 500
|
*/
app()->setErrorHandler(function ($e = null) {
    if ($e) {
        if (app()->config("log.enabled")) {
            app()->logger()->error($e);
        }
    }

    response()->json("An error occured, our team has been notified", 500, true);
});

/*
|--------------------------------------------------------------------------
| Set up Controller namespace
|--------------------------------------------------------------------------
|
| This allows you to directly use controller names instead of typing
| the controller namespace first.
|
*/
app()->setNamespace("\App\Controllers");

// You can break up routes into individual files
require __DIR__ . "/_app.php";


```

**Reply by fadrian06** [reply]
I'll try to reproduce the error...

When you say it doesn't work? Do you mean that the route gives 404, or some exception or error thrown?

**Comment by maswebber** [comment]
Hello, I'm new to Leaf PHP API, 
I tried to use auth module, i did the auth:scaffold, and then trying to login, but it keeps saying this 

`Leaf\Exceptions\ErrorException thrown with message "You need to connect to your database first"`

so i tried, manually connect to database using model, and its working fine, i could get user data, easily, but `auth->login` seems not connected to database, the user model i use is default user model.

**Reply by maswebber** [reply]
Hmm i tried adding `auth()->autoConnect()` inside `/root/config/auth.php` and its working now, but is it the best practice ?

**Reply by mychidarko** [reply]
You should rather head over to the public/index.php file and uncomment the initDb line

**Reply by maswebber** [reply]
oh shoot, never mind, after i read inside `root/public/index.php` i need to uncomment `\Leaf\Database::initDb();`

**Comment by maswebber** [comment]
> You should rather head over to the public/index.php file and uncomment the initDb line

Thank you

**Comment by maswebber** [comment]
Hello,
I tried to validate `item_desc` attribute that optional, so i tried something like image below :
![image](https://github.com/leafsphp/leaf/assets/60123041/733f256a-af0b-43de-9607-bf195ff3dcd3)
but i got message like this:
![image](https://github.com/leafsphp/leaf/assets/60123041/1065a6ae-54c5-49a5-a61e-81e5676ee6b7)
the attribute doesn't have the required rule, but it still required. is it intended work like that ? if yes then i will using custom rule

**Reply by maswebber** [reply]
i use this custom rule, with allow empty field and if text exist check the text with alphanum regex, love the framework 👍
![image](https://github.com/leafsphp/leaf/assets/60123041/457d30e9-4443-4be0-8b18-ffc19c12ff75)

**Reply by fadrian06** [reply]
I seem to remember that before you had to add the optional rule

```diff
- => 'text'
+ => 'optional|text'
```

**Comment by maswebber** [comment]
Form validation not valid json

Input Data: 
`addons_data : [{"price_id": 4}, {"price_id": 5}]` //This doesn't work
`addons_data : ["test"]` //This work when $safedata false
Both valid json

Controller
Get Data:
`$newData=request()->get(['addons_data'],false);`

Validation:
`$validation=form()->validate($newData,['addons_data'=>['json']]);
 if(!$validation){
response()->exit(form()->errors());
}`

Any idea how to fix this ?

**Comment by EnMaster** [comment]
Hi, is it possible to automatically generate multiple JSON schemas from an existing MySQL database?

**Reply by EnMaster** [reply]
Ok, i've found a solution, but using laravel with this tool: [laravel-migrations-generator](https://github.com/kitloong/laravel-migrations-generator/tree/6.x). Then, I copy the files by hand and make them compatible for LeafPHP. However, it would be nice to have a similar tool for LeafPHP.

**Comment by maswebber** [comment]
How to use DB::transaction() in LeafPHP ?

**Reply by mychidarko** [reply]
If you are using Leaf DB, it does not yet come with a `transaction()` method

**Reply by fadrian06** [reply]
@mychidarko I could help you implement it, couldn't I?

**Reply by mychidarko** [reply]
Absolutely!

**Comment by franklinvillalobos** [comment]
Does anyone know how I can add tailwindcss to leaf 3 mvc optimally? When I have tried to do it, some errors always come out.

**Reply by mychidarko** [reply]
Hi @franklinvillalobos can you share what you already tried?

**Reply by mychidarko** [reply]
Also, the Leaf MVC's console comes with a `view:install` command. You can add tailwind like this:

```php
php leaf view:install --tailwind
```

**Reply by franklinvillalobos** [reply]
Thank you! I was able to solve it, I had to modify some lines of the core for it to work correctly.

**Comment by franklinvillalobos** [comment]
Hello, I want to know how I can translate leaf's error messages into another language.
Where is the file to add or change the language?

**Reply by fadrian06** [reply]
If you are referring to the validation messages For example, you could configure the messages with

```php
\Leaf\Form::message([
  'required' => '{field} is required',
  'text' => 'your custom message...',
  /*...*/
]);
```

**Reply by franklinvillalobos** [reply]
But from here I don't see how to translate the auth error messages, for example the errors issued by auth()->errors();

**Comment by WebworkStudios** [comment]
Can I use ADR patterns with Leaf?
I like __invoke Actions.

**Comment by senthiece42** [comment]
<img width="802" alt="Screenshot 2024-01-13 at 11 40 17 PM" src="https://github.com/leafsphp/leaf/assets/59682893/3d3f6ece-a5b1-4481-a1dd-707d21324b23">

I had tried to migrate (php leaf db:migrate) cmd for PostgreSQL db while I  am getting (refer above image) those kind of issues.. and not properly working as well MySQL side also the db:migration cmds getting some different kind of issue.. could pls look into that..

**Comment by 7Million** [comment]
This is a great initiative, it is not too-loaded like other major frameworks. However, the documentation is quite confusing and too linked. I am struggling to understand how I can integrate Leaf API with Leaf UI, I am trying to transition from Laravel + Livewire but the docs are taking me around.

**Comment by tahacicek** [comment]
I think I put the 1,000th star in.

**Comment by romariotech** [comment]
How to configure config/database.php to connect to different banks? It's possible?
